In [1]:
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.preprocessing import LabelEncoder

In [2]:
train_df = pd.read_csv(Path('Resources/2019loans.csv'))
test_df = pd.read_csv(Path('Resources/2020Q1loans.csv'))

In [3]:
# Define X_train and y_train
y_train_1 = train_df["loan_status"]
X_train_1 = train_df.drop(columns = ["loan_status"])

In [4]:
# Define X_test and y_test
y_test_1 = test_df["loan_status"]
X_test_1 = test_df.drop(columns = ["loan_status"])

In [5]:
# Convert categorical data to numeric and separate target feature for testing data and training data
y_train = LabelEncoder().fit_transform(train_df['loan_status'])
y_test = LabelEncoder().fit_transform(test_df['loan_status'])
X_train = pd.get_dummies(X_train_1)
X_test = pd.get_dummies(X_test_1)

In [6]:
# add missing dummy variables to testing set
missing_cols = set( X_train.columns ) - set( X_test.columns )
# Add a missing column in test set with default value equal to 0
for c in missing_cols:
    X_test[c] = 0
# Ensure the order of column in the test set is in the same order than in train set
X_test = X_test[X_train.columns]

In [7]:
# Train the Logistic Regression model on the unscaled data and print the model score
from sklearn.linear_model import LogisticRegression
clf_lr = LogisticRegression()
clf_lr.fit(X_train, y_train)

print(f"Training Data Score: {clf_lr.score(X_train, y_train)}")
print(f"Testing Data Score: {clf_lr.score(X_test, y_test)}")

Training Data Score: 0.648440065681445
Testing Data Score: 0.5253083794130158


/Users/negindjalali/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [8]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier

clf = RandomForestClassifier(random_state=1, n_estimators=500).fit(X_train, y_train)
print(f'Training Score: {clf.score(X_train, y_train)}')
print(f'Testing Score: {clf.score(X_test, y_test)}')

Training Score: 1.0
Testing Score: 0.6180348787749894


In [9]:
# Scale the data
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.fit_transform(X_test)

In [10]:
# Train the Logistic Regression model on the scaled data and print the model score
clf_lr.fit(X_train_scaled, y_train)
print(f"Training Data Score: {clf_lr.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {clf_lr.score(X_test_scaled, y_test)}")

Training Data Score: 0.7132183908045977
Testing Data Score: 0.670565716716291


/Users/negindjalali/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [11]:
# Train a Random Forest Classifier model on the scaled data and print the model score
clf = RandomForestClassifier(random_state=1, n_estimators=500).fit(X_train_scaled, y_train)
print(f'Training Score: {clf.score(X_train_scaled, y_train)}')
print(f'Testing Score: {clf.score(X_test_scaled, y_test)}')


Training Score: 1.0
Testing Score: 0.5854955338153977


Analysis:
Based on our data analysis the random forest classifier without scaled data resulted in training score of 1.0 and testing score of 0.46 which is comparable to the random forest classifer with scaled data which resulted in training score of 1.0 and testing score of 0.59.
However the model for linear regression works better for the scaled data than unscaled data.